File: scripts/diag_summary.py
#!/usr/bin/env python3
"""Aggregate diag/<bench>/stats.json files + benchmarks/data/reference.json
into diag/summary.md for Step 1 of a round. Stdout is the summary markdown.

Written to be called from scripts/diag.sh with diag/ as the argument.
"""

import json
import re
import sys
from pathlib import Path


def parse_time(value):
    """Reference/latest JSON records 'jit' as '0.085s' or similar. Return
    float seconds or None."""
    if value is None:
        return None
    m = re.search(r"[\d.]+", str(value))
    return float(m.group()) if m else None


def main(diag_root):
    diag_root = Path(diag_root)
    if not diag_root.is_dir():
        print(f"error: {diag_root} is not a directory", file=sys.stderr)
        sys.exit(2)

    ref_path = Path("benchmarks/data/reference.json")
    ref = None
    if ref_path.exists():
        with open(ref_path) as f:
            ref = json.load(f)

    latest_path = Path("benchmarks/data/latest.json")
    latest = None
    if latest_path.exists():
        with open(latest_path) as f:
            latest = json.load(f)

    # Layout v2: diag/<domain>/<bench>/stats.json.
    # Layout v1 (legacy): diag/<bench>/stats.json — still tolerated.
    benches = []  # list of (label, domain, name, stats_path)
    for top in sorted(diag_root.iterdir()):
        if not top.is_dir():
            continue
        if (top / "stats.json").exists():
            benches.append((top.name, "legacy", top.name, top))
            continue
        for sub in sorted(top.iterdir()):
            if sub.is_dir() and (sub / "stats.json").exists():
                benches.append((f"{top.name}/{sub.name}", top.name, sub.name, sub))

    lines = []
    lines.append("# diag/summary.md")
    lines.append("")
    lines.append(f"Generated across {len(benches)} benchmarks.")
    lines.append("")
    lines.append("## Insn counts (hottest proto per benchmark)")
    lines.append("")
    lines.append("| Benchmark | Domain | Hottest proto | Insns | Bytes | Load | Store | FP | Branch |")
    lines.append("|-----------|-------|---------------|------:|------:|-----:|------:|---:|-------:|")

    per_bench_hottest = {}
    for label, domain, name, bench_dir in benches:
        with open(bench_dir / "stats.json") as f:
            data = json.load(f)
        protos = [p for p in data.get("protos", []) if not p.get("skip_reason")]
        if not protos:
            continue
        hot = max(protos, key=lambda p: p.get("insn_count", 0))
        hist = hot.get("insn_histogram", {})
        per_bench_hottest[label] = hot
        lines.append("| {b} | {s} | {n} | {i} | {bytes} | {l} | {st} | {f} | {br} |".format(
            b=name,
            s=domain,
            n=hot.get("name", "?"),
            i=hot.get("insn_count", 0),
            bytes=hot.get("code_bytes", 0),
            l=hist.get("load", 0),
            st=hist.get("store", 0),
            f=hist.get("fp", 0),
            br=hist.get("branch", 0),
        ))

    if ref and latest:
        lines.append("")
        lines.append("## Drift vs frozen reference.json")
        lines.append("")
        ref_meta = ref.get("_meta", {})
        excluded = set(ref_meta.get("excluded", []))
        flag_t = ref_meta.get("drift_threshold_flag_pct", 2.0)
        fail_t = ref_meta.get("drift_threshold_fail_pct", 5.0)

        drifters = []
        ref_results = ref.get("results", {})
        latest_results = latest.get("results", {})
        for k, v in ref_results.items():
            if k in excluded:
                continue
            rj = parse_time(v.get("jit"))
            lj = parse_time(latest_results.get(k, {}).get("jit"))
            if rj and lj and rj > 0:
                drifters.append((k, rj, lj, (lj - rj) / rj * 100))
        drifters.sort(key=lambda x: -x[3])

        lines.append(f"Flag threshold: {flag_t}%. Fail threshold: {fail_t}%.")
        lines.append("")
        lines.append("| Benchmark | Reference | Latest | Drift | Status |")
        lines.append("|-----------|----------:|-------:|------:|:-------|")
        for k, r, l, d in drifters[:10]:
            status = "FAIL" if d >= fail_t else ("FLAG" if d >= flag_t else "ok")
            lines.append(f"| {k} | {r:.3f}s | {l:.3f}s | {d:+.2f}% | {status} |")
    else:
        lines.append("")
        lines.append("## Drift vs frozen reference.json")
        lines.append("")
        if not ref:
            lines.append("_reference.json not found_")
        if not latest:
            lines.append("_latest.json not found — run benchmarks/strict_guard.py or benchmarks/regression_guard.py first_")

    lines.append("")
    lines.append("## Histogram anomalies")
    lines.append("")
    anomalies = []
    for bench, hot in per_bench_hottest.items():
        hist = hot.get("insn_histogram", {})
        loads = hist.get("load", 0)
        stores = hist.get("store", 0)
        total = hot.get("insn_count", 1)
        mem_pct = 100 * (loads + stores) / total if total else 0
        if mem_pct > 50:
            anomalies.append(
                f"- **{bench}/{hot.get('name','?')}**: {loads}+{stores}={loads+stores} memory ops "
                f"out of {total} insns ({mem_pct:.0f}%) — dominant store/load, suspect allocation/field access."
            )
    if anomalies:
        lines.extend(anomalies)
    else:
        lines.append("_no benchmarks above 50% memory-op threshold._")

    print("\n".join(lines))


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("usage: diag_summary.py <diag_root>", file=sys.stderr)
        sys.exit(2)
    main(sys.argv[1])

File: cmd/leia/diag_summary.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

type diagSummaryBench struct {
	Label  string
	Domain string
	Name   string
	Dir    string
}

type diagStatsFile struct {
	Protos []diagProtoStats `json:"protos"`
}

type diagProtoStats struct {
	Name          string         `json:"name"`
	SkipReason    string         `json:"skip_reason"`
	InsnCount     int            `json:"insn_count"`
	CodeBytes     int            `json:"code_bytes"`
	InsnHistogram map[string]int `json:"insn_histogram"`
}

func runDiagSummaryCommand(args []string, outw, errw io.Writer) int {
	fs := flag.NewFlagSet("diag summary", flag.ContinueOnError)
	fs.SetOutput(errw)
	if err := fs.Parse(args); err != nil {
		if err == flag.ErrHelp {
			return 0
		}
		return 2
	}
	if len(fs.Args()) != 1 {
		fmt.Fprintln(errw, "usage: leia diag summary <diag-root>")
		return 2
	}
	text, err := buildDiagSummary(fs.Args()[0])
	if err != nil {
		fmt.Fprintf(errw, "leia diag summary: %v\n", err)
		return 1
	}
	fmt.Fprint(outw, text)
	if !strings.HasSuffix(text, "\n") {
		fmt.Fprintln(outw)
	}
	return 0
}

func buildDiagSummary(diagRoot string) (string, error) {
	info, err := os.Stat(diagRoot)
	if err != nil {
		return "", err
	}
	if !info.IsDir() {
		return "", fmt.Errorf("%s is not a directory", diagRoot)
	}
	benches, err := discoverDiagSummaryBenches(diagRoot)
	if err != nil {
		return "", err
	}

	var b strings.Builder
	fmt.Fprintln(&b, "# diag/summary.md")
	fmt.Fprintln(&b)
	fmt.Fprintf(&b, "Generated across %d benchmarks.\n", len(benches))
	fmt.Fprintln(&b)
	fmt.Fprintln(&b, "## Insn counts (hottest proto per benchmark)")
	fmt.Fprintln(&b)
	fmt.Fprintln(&b, "| Benchmark | Domain | Hottest proto | Insns | Bytes | Load | Store | FP | Branch |")
	fmt.Fprintln(&b, "|-----------|-------|---------------|------:|------:|-----:|------:|---:|-------:|")

	hottest := map[string]diagProtoStats{}
	for _, bench := range benches {
		stats, err := readDiagStats(filepath.Join(bench.Dir, "stats.json"))
		if err != nil {
			return "", err
		}
		hot, ok := hottestProto(stats.Protos)
		if !ok {
			continue
		}
		hottest[bench.Label] = hot
		hist := hot.InsnHistogram
		fmt.Fprintf(
			&b,
			"| %s | %s | %s | %d | %d | %d | %d | %d | %d |\n",
			bench.Name,
			bench.Domain,
			emptyAsQuestion(hot.Name),
			hot.InsnCount,
			hot.CodeBytes,
			hist["load"],
			hist["store"],
			hist["fp"],
			hist["branch"],
		)
	}

	writeDiagDriftSummary(&b)
	fmt.Fprintln(&b)
	fmt.Fprintln(&b, "## Histogram anomalies")
	fmt.Fprintln(&b)
	anomalies := make([]string, 0)
	labels := make([]string, 0, len(hottest))
	for label := range hottest {
		labels = append(labels, label)
	}
	sort.Strings(labels)
	for _, label := range labels {
		hot := hottest[label]
		hist := hot.InsnHistogram
		loads := hist["load"]
		stores := hist["store"]
		total := hot.InsnCount
		if total == 0 {
			total = 1
		}
		memPct := 100 * float64(loads+stores) / float64(total)
		if memPct > 50 {
			anomalies = append(anomalies, fmt.Sprintf(
				"- **%s/%s**: %d+%d=%d memory ops out of %d insns (%.0f%%) — dominant store/load, suspect allocation/field access.",
				label,
				emptyAsQuestion(hot.Name),
				loads,
				stores,
				loads+stores,
				total,
				memPct,
			))
		}
	}
	if len(anomalies) == 0 {
		fmt.Fprintln(&b, "_no benchmarks above 50% memory-op threshold._")
	} else {
		for _, line := range anomalies {
			fmt.Fprintln(&b, line)
		}
	}
	return b.String(), nil
}

func discoverDiagSummaryBenches(diagRoot string) ([]diagSummaryBench, error) {
	entries, err := os.ReadDir(diagRoot)
	if err != nil {
		return nil, err
	}
	var benches []diagSummaryBench
	for _, top := range entries {
		if !top.IsDir() {
			continue
		}
		topDir := filepath.Join(diagRoot, top.Name())
		if fileExists(filepath.Join(topDir, "stats.json")) {
			benches = append(benches, diagSummaryBench{Label: top.Name(), Domain: "previous_schema", Name: top.Name(), Dir: topDir})
			continue
		}
		subs, err := os.ReadDir(topDir)
		if err != nil {
			return nil, err
		}
		for _, sub := range subs {
			if !sub.IsDir() {
				continue
			}
			subDir := filepath.Join(topDir, sub.Name())
			if fileExists(filepath.Join(subDir, "stats.json")) {
				benches = append(benches, diagSummaryBench{Label: top.Name() + "/" + sub.Name(), Domain: top.Name(), Name: sub.Name(), Dir: subDir})
			}
		}
	}
	sort.Slice(benches, func(i, j int) bool { return benches[i].Label < benches[j].Label })
	return benches, nil
}

func readDiagStats(path string) (diagStatsFile, error) {
	var stats diagStatsFile
	data, err := os.ReadFile(path)
	if err != nil {
		return stats, err
	}
	err = json.Unmarshal(data, &stats)
	return stats, err
}

func hottestProto(protos []diagProtoStats) (diagProtoStats, bool) {
	var hot diagProtoStats
	ok := false
	for _, proto := range protos {
		if proto.SkipReason != "" {
			continue
		}
		if !ok || proto.InsnCount > hot.InsnCount {
			hot = proto
			ok = true
		}
	}
	if hot.InsnHistogram == nil {
		hot.InsnHistogram = map[string]int{}
	}
	return hot, ok
}

func writeDiagDriftSummary(b *strings.Builder) {
	fmt.Fprintln(b)
	fmt.Fprintln(b, "## Drift vs frozen reference.json")
	fmt.Fprintln(b)
	ref, refOK := readBenchmarkTimingJSON(filepath.Join("benchmarks", "data", "reference.json"))
	latest, latestOK := readBenchmarkTimingJSON(filepath.Join("benchmarks", "data", "latest.json"))
	if !refOK || !latestOK {
		if !refOK {
			fmt.Fprintln(b, "_reference.json not found_")
		}
		if !latestOK {
			fmt.Fprintln(b, "_latest.json not found — run benchmarks/strict_guard.py or benchmarks/regression_guard.py first_")
		}
		return
	}
	meta := mapValue(ref["_meta"])
	excluded := map[string]bool{}
	if values, ok := meta["excluded"].([]any); ok {
		for _, value := range values {
			excluded[fmt.Sprint(value)] = true
		}
	}
	flagThreshold := floatFromAny(meta["drift_threshold_flag_pct"], 2.0)
	failThreshold := floatFromAny(meta["drift_threshold_fail_pct"], 5.0)
	refResults := mapValue(ref["results"])
	latestResults := mapValue(latest["results"])
	type drifter struct {
		Name string
		Ref  float64
		New  float64
		Pct  float64
	}
	var drifters []drifter
	for name, rawRef := range refResults {
		if excluded[name] {
			continue
		}
		refTime, ok := parseDiagTime(mapValue(rawRef)["jit"])
		if !ok || refTime <= 0 {
			continue
		}
		latestTime, ok := parseDiagTime(mapValue(latestResults[name])["jit"])
		if !ok {
			continue
		}
		drifters = append(drifters, drifter{Name: name, Ref: refTime, New: latestTime, Pct: (latestTime - refTime) / refTime * 100})
	}
	sort.Slice(drifters, func(i, j int) bool { return drifters[i].Pct > drifters[j].Pct })
	fmt.Fprintf(b, "Flag threshold: %.1f%%. Fail threshold: %.1f%%.\n", flagThreshold, failThreshold)
	fmt.Fprintln(b)
	fmt.Fprintln(b, "| Benchmark | Reference | Latest | Drift | Status |")
	fmt.Fprintln(b, "|-----------|----------:|-------:|------:|:-------|")
	for i, d := range drifters {
		if i >= 10 {
			break
		}
		status := "ok"
		if d.Pct >= failThreshold {
			status = "FAIL"
		} else if d.Pct >= flagThreshold {
			status = "FLAG"
		}
		fmt.Fprintf(b, "| %s | %.3fs | %.3fs | %+.2f%% | %s |\n", d.Name, d.Ref, d.New, d.Pct, status)
	}
}

func readBenchmarkTimingJSON(path string) (map[string]any, bool) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, false
	}
	var out map[string]any
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, false
	}
	return out, true
}

var diagTimeRE = regexp.MustCompile(`[\d.]+`)

func parseDiagTime(value any) (float64, bool) {
	if value == nil {
		return 0, false
	}
	match := diagTimeRE.FindString(fmt.Sprint(value))
	if match == "" {
		return 0, false
	}
	parsed, err := strconv.ParseFloat(match, 64)
	return parsed, err == nil
}

func mapValue(value any) map[string]any {
	if out, ok := value.(map[string]any); ok {
		return out
	}
	return map[string]any{}
}

func floatFromAny(value any, fallback float64) float64 {
	switch v := value.(type) {
	case float64:
		return v
	case int:
		return float64(v)
	default:
		return fallback
	}
}

func emptyAsQuestion(value string) string {
	if value == "" {
		return "?"
	}
	return value
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}
